Dockerfile

```diff
@@ -1,24 +1,29 @@
+# Stage 1: Precompile TestRunner and Module Cache
 FROM swift:6.1.2 AS builder
-# WORKDIR /opt/testrunner
-
-COPY src/TestRunner ./
-
-# Print Installed Swift Version
 RUN swift --version
-#RUN swift package clean
+
+# Build TestRunner executable
+WORKDIR /TestRunner
+COPY src/TestRunner .
 RUN swift build --configuration release
 
+# Build WarmUp package
+# Build directory and final working paths should be equal for reuse of ModuleCache.
+WORKDIR /opt/test-runner
+COPY src/WarmUp .
+RUN swift build --build-tests
+
+# Stage 2: Prepare docker container image
 FROM swift:6.1.2
 RUN apt-get update && apt-get install -y jq
-WORKDIR /opt/test-runner/
-COPY bin/ bin/
-
-COPY Package.swift ./Package.swift
 
-RUN swift build
-
-COPY --from=builder /.build/release/TestRunner bin/
+WORKDIR /opt/test-runner/
+COPY bin/run.sh bin/run.sh
+COPY bin/run-test.sh bin/run-test.sh
+COPY --from=builder /TestRunner/.build/release/TestRunner bin/
+COPY --from=builder /opt/test-runner/.build .build
+COPY --from=builder /opt/test-runner/Package.resolved Package.resolved
```

Comment:
You may want to combine these to reduce the number of layers.

Reply:
Not sure it's possible here. Each copy operation has it's own destination: TestRunner and Pacage.resolved are files, .build is a directory.